src/core/Akka.Streams/Implementation/Fusing/StreamOfStreams.cs

```diff
@@ -1,576 +1,809 @@
+            if (pub != null)
+            {
+                Unit _;
+                pub.Create(default(MaterializationContext), out _).Subscribe(CancelingSubscriber<T>.Instance);
+                return;
+            }
+
+            var intp = GraphInterpreter.CurrentInterpreterOrNull;
+            if (intp == null)
+                throw new NotSupportedException($"cannot drop Source of type {s.Module.GetType().Name}");
+            s.RunWith(Sink.Ignore<T>(), intp.SubFusingMaterializer);
+        }
+    }
+
+    /// <summary>
+    /// INTERNAL API
+    /// </summary>
+    internal sealed class SubSource<T> : GraphStage<SourceShape<T>>
+    {
+        #region internal classes 
+
+        private sealed class SubSourceLogic : GraphStageLogic
+        {
+            private readonly SubSource<T> _sub;
+
+            public SubSourceLogic(SubSource<T> sub) : base(sub.Shape)
+            {
+                _sub = sub;
+
+                SetHandler(sub._out, onPull: () => sub.ExternalCallback(SubSink.Command.RequestOne), onDownstreamFinish: () => sub.ExternalCallback(SubSink.Command.Cancel));
+            }
+
+            private void SetCb(Action<IActorSubscriberMessage> cb)
+            {
+                var s = _sub._status.Value;
+
+                if (s == null)
+                {
+                    if (!_sub._status.CompareAndSet(null, cb))
+                        SetCb(cb);
+                    return;
+                }
+
+                var complete = s as OnComplete;
+                if (complete != null)
+                {
+                    CompleteStage();
+                    return;
+                }
+
+                var error = s as OnError;
+                if (error != null)
+                {
+                    FailStage(error.Cause);
+                    return;
                 }
-                else FailStage(cause);
+
+                if (s is Action<IActorSubscriberMessage>)
+                    throw new IllegalStateException("Substream Source cannot be materialized more than once");
             }
 
-            private void OnDownstreamFinish()
+            public override void PreStart()
             {
-                if(!IsPrefixComplete) CompleteStage();
-                // Otherwise substream is open, ignore
+                SetCb(message => { message.Match().With<OnComplete>(CompleteStage).With<OnError>(e => FailStage(e.Cause)).With<OnNext>(n => Push(_sub._out, (T) n.Element)); });
             }
         }
 
         #endregion
 
-        public readonly int Count;
-        public readonly Inlet<T> In = new Inlet<T>("PrefixAndTail.in");
-        public readonly Outlet<Tuple<IEnumerable<T>, Source<T, Unit>>> Out = new Outlet<Tuple<IEnumerable<T>, Source<T, Unit>>>("PrefixAndTail.out");
+        private readonly string _name;
+        private readonly Outlet<T> _out = new Outlet<T>("SubSource.out");
+        private readonly AtomicReference<object> _status = new AtomicReference<object>();
```

Comment:
It looks like this container always have values of type `Action<IActorSubscriberMessage>`. Shouldn't we specify generic param to that type?